sklearn/ensemble/forest.py

```diff
@@ -72,17 +73,56 @@ class calls the ``fit`` method of each sub-estimator on random samples
 MAX_INT = np.iinfo(np.int32).max
 
 
-def _generate_sample_indices(random_state, n_samples):
+def _get_n_samples_bootstrap(n_samples, max_samples):
+    """Get the number of samples in a bootstrap sample.
+
+    Parameters
+    ----------
+    n_samples : int
+        Number of samples in the dataset.
+    max_samples : int or float
+        The maximum number of samples to draw from the total available:
+            - if float, this indicates a fraction of the total;
+            - if int, this indicates the exact number of samples;
+            - if None, this indicates the total number of samples.
+
+    Returns
+    -------
+    n_samples_bootstrap : int
+        The total number of samples to draw for the bootstrap sample.
+    """
+    if max_samples is None:
+        return n_samples
+
+    if isinstance(max_samples, numbers.Integral):
```

Comment:
I think we should be consistent w.r.t how we treat these fractions. For instance, in optics, we have:

https://github.com/scikit-learn/scikit-learn/blob/e5be3ad025fe372f4811a2c5ffb5b6fb5bfcc8f1/sklearn/cluster/optics_.py#L280-L290

And then `1` always means `100%` of the data, at least in optics. Do we have a similar semantics in other places?

Reply:
With PCA, `n_components=1` means 1 components while `n_components<1` will be a percentage.

Reply:
Excluding `1` from float avoid issue with `float` comparison as well.